### PR TITLE
First pass at porting the UX interactive controls

### PR DESCRIPTION
**Pull request by SimonDarksideJ**
Overview
---
First pass at porting the UX interactive controls and associated features.

TODO
---
- [ ] Resolve Click Handler
- [ ] Resolve Gesture Input
- [ ] Resolve Speech Input

Changes
---
- Adds new InteractiveTheme Scriptable configuration (not implemented in interactables yet)
- Adds Core interactive Controls
- Adds related Interactive Widgets
- Adds some new base utilities for Move / Scale / Fade and ColorFade

### CODE REVIEW

File: Assets/MixedRealityToolkit/_Core/Utilities/ColorTransition.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            {
+                ColorTransitionData data = new ColorTransitionData();
+                data.StartColor = Materials[i].color;
+                data.Percentage = 0;
+                data.Time = TransitionTime;
+                data.Count = TransitionTime;
+                data.Material = Materials[i];
+                data.Name = Materials[i].name;
+
+                int SpaceIndex = data.Name.IndexOf(" ");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Utilities/ColorTransition.cs]: just white space?

File: Assets/MixedRealityToolkit/_Core/Utilities/ColorTransition.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            {
+                ColorTransitionData data = mData[i];
+
+                if (data.Count < data.Time)
+                {
+                    data.Count = Mathf.Clamp(data.Count + Time.deltaTime, 0, data.Time);
+                    data.Material.color = GetColorTransition(data.StartColor, data.EndColor, data.Count / data.Time);
+                    mData[i] = data;
+                }
+
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Utilities/ColorTransition.cs]: extra line

File: Assets/MixedRealityToolkit/_Core/Utilities/FadeColors.cs [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+        public LerpTypes LerpType;
+
+        [Tooltip("Time to fade in seconds")]
+        public float LerpTime = 1f;
+
+        [Tooltip("Run by default? or Status")]
+        public bool IsRunning = false;
+
+        [Tooltip("Animation is complete!")]
+        public UnityEvent OnComplete;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Utilities/FadeColors.cs]: Why? We have our own event system to raise events on.

File: Assets/MixedRealityToolkit/_Core/Utilities/FadeColors.cs [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        /// The selectable ease types for the animation
+        ///     Linear: steady progress
+        ///     EaseIn: ramp up in speed
+        ///     EaseOut: ramp down in speed
+        ///     EaseInOut: ramp up then down in speed
+        /// </summary>
+        public enum LerpTypes { Linear, EaseIn, EaseOut, EaseInOut }
+
+        [Tooltip("Fade ease type")]
+        public GameObject TargetObject;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Utilities/FadeColors.cs]: public fields should be made private with public property accessors if needed.

File: Assets/MixedRealityToolkit/_Core/Utilities/FadeColors.cs [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+        // animation ticker
+        private float mLerpTimeCounter;
+
+        // color and alpha values
+        private Color mCachedColor;
+        private float mCurrentAlpha;
+
+        // component targets
+        private Material mMaterial;
+        private TextMesh mTextMesh;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Utilities/FadeColors.cs]: Why TextMesh? Let's use TextMeshPro.

File: Assets/MixedRealityToolkit/_Core/Utilities/FadeColors.cs [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        {
+            if (mTextMesh != null)
+            {
+                mCachedColor = mTextMesh.color;
+            }
+            else
+            {
+                if (mMaterial != null)
+                {
+                    mCachedColor = mMaterial.color;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Utilities/FadeColors.cs]: memory leak.

File: Assets/MixedRealityToolkit/_Core/Utilities/FadeColors.cs [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        // color and alpha values
+        private Color mCachedColor;
+        private float mCurrentAlpha;
+
+        // component targets
+        private Material mMaterial;
+        private TextMesh mTextMesh;
+
+        // are we fading in or out?
+        private bool mIsFadingIn = true;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Utilities/FadeColors.cs]: Let's also update the names of these to comply with our usual standard. (drop the m prefixes)

**Comment by cre8ivepark** [comment]
@killerantz Todd, shouldn't we add your new button system components instead of this one?

**Comment by StephenHodgson** [comment]
Yeah that's a very good question

**Comment by StephenHodgson** [comment]
Let's hold off on this port until Todd reworks this feature.

**Comment by SimonDarksideJ** [comment]
I agree, this is only here to promote discussion about what to keep and what to rework.
@killerantz if you have something ready to go, then this should be pulled in favour of yours

**Comment by killerantz** [comment]
It would be great to get any of the work you have done so far I can easily integrate what I have, which is using the current MRTK input system. But the other button features are getting reworked and should get moved over when we switch to Alpha. I have most of the code framed in and I'm starting to recreate all the UI components and samples now. The outline was posted in issue #1625

**Comment by SimonDarksideJ** [comment]
I'll leave the branch live @killerantz so you can grab / review what you like.  Key thing is to use the new Input System Handler interfaces to react to input and the actions defined by them.

Only part of the code I'd like to see kept was the:
"Assets/MixedRealityToolkit/_Core/Definitions/UX/InteractiveTheme.cs" code as it worked extremely well as a scriptable object.  Allowing reusable configurable profiles for types of themes.
Where possible, we should use profiles for theme style settings, rather than forcing the user to visit each one individually.

**Pull request closed without merging**